src/Microsoft.Net.Sdk.AnalyzerRedirecting/SdkAnalyzerAssemblyRedirector.cs

```diff
@@ -0,0 +1,132 @@
+
+        // Expects layout like:
+        // VsInstallDir\SDK\RuntimeAnalyzers\WindowsDesktopAnalyzers\8.0.8\analyzers\dotnet\System.Windows.Forms.Analyzers.dll
+        //                                   ~~~~~~~~~~~~~~~~~~~~~~~                                                           = topLevelDirectory
+        //                                                           ~~~~~                                                     = versionDirectory
+        //                                                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ = analyzerPath
+
+        foreach (string topLevelDirectory in Directory.EnumerateDirectories(_insertedAnalyzersDirectory))
+        {
+            foreach (string versionDirectory in Directory.EnumerateDirectories(topLevelDirectory))
+            {
+                foreach (string analyzerPath in Directory.EnumerateFiles(versionDirectory, "*.dll", SearchOption.AllDirectories))
+                {
+                    if (!analyzerPath.StartsWith(versionDirectory, StringComparison.OrdinalIgnoreCase))
+                    {
+                        continue;
+                    }
+
+                    string version = Path.GetFileName(versionDirectory);
+                    string analyzerName = Path.GetFileNameWithoutExtension(analyzerPath);
+                    string pathSuffix = analyzerPath.Substring(versionDirectory.Length + 1 /* slash */);
+                    pathSuffix = Path.GetDirectoryName(pathSuffix);
+
+                    AnalyzerInfo analyzer = new() { FullPath = analyzerPath, ProductVersion = version, PathSuffix = pathSuffix };
+
+                    if (builder.TryGetValue(analyzerName, out var existing))
+                    {
+                        existing.Add(analyzer);
+                    }
+                    else
+                    {
+                        builder.Add(analyzerName, [analyzer]);
+                    }
+                }
+            }
+        }
+
+        return builder.ToImmutable();
+    }
+
+    public string? RedirectPath(string fullPath)
+    {
+        if (AnalyzerMap.TryGetValue(Path.GetFileNameWithoutExtension(fullPath), out var analyzers))
+        {
+            foreach (AnalyzerInfo analyzer in analyzers)
+            {
+                var directoryPath = Path.GetDirectoryName(fullPath);
+                if (directoryPath.EndsWith(analyzer.PathSuffix, StringComparison.OrdinalIgnoreCase) &&
+                    majorAndMinorVersionsMatch(directoryPath, analyzer.PathSuffix, analyzer.ProductVersion))
+                {
+                    return analyzer.FullPath;
+                }
+            }
+        }
+
+        return null;
+
+        static bool majorAndMinorVersionsMatch(string directoryPath, string pathSuffix, string version)
+        {
+            // Find the version number in the directory path - it is in the directory name before the path suffix.
```

Comment:
It might be worth calling out that this can also come from a  NuGet package which would have the same version folder in the path.